Source/Core/Core/IOS/USB/Emulated/Skylander.cpp

```diff
@@ -0,0 +1,825 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/IOS/USB/Emulated/Skylander.h"
+
+#include <map>
+#include <mutex>
+#include <vector>
+
+#include "Common/Logging/Log.h"
+#include "Common/StringUtil.h"
+#include "Common/Timer.h"
+#include "Core/Core.h"
+#include "Core/HW/Memmap.h"
+#include "Core/System.h"
+
+namespace IOS::HLE::USB
+{
+SkylanderUSB::SkylanderUSB(Kernel& ios, const std::string& device_name) : m_ios(ios)
+{
+  m_vid = 0x1430;
+  m_pid = 0x150;
+  m_id = (static_cast<u64>(m_vid) << 32 | static_cast<u64>(m_pid) << 16 | static_cast<u64>(9) << 8 |
+          static_cast<u64>(1));
+  m_device_descriptor = DeviceDescriptor{0x12,   0x1,   0x200, 0x0, 0x0, 0x0, 0x40,
+                                         0x1430, 0x150, 0x100, 0x1, 0x2, 0x0, 0x1};
+  m_config_descriptor.emplace_back(ConfigDescriptor{0x9, 0x2, 0x29, 0x1, 0x1, 0x0, 0x80, 0xFA});
+  m_interface_descriptor.emplace_back(
+      InterfaceDescriptor{0x9, 0x4, 0x0, 0x0, 0x2, 0x3, 0x0, 0x0, 0x0});
+  m_endpoint_descriptor.emplace_back(EndpointDescriptor{0x7, 0x5, 0x81, 0x3, 0x40, 0x1});
+  m_endpoint_descriptor.emplace_back(EndpointDescriptor{0x7, 0x5, 0x2, 0x3, 0x40, 0x1});
+}
+
+SkylanderUSB::~SkylanderUSB() = default;
+
+DeviceDescriptor SkylanderUSB::GetDeviceDescriptor() const
+{
+  return m_device_descriptor;
+}
+
+std::vector<ConfigDescriptor> SkylanderUSB::GetConfigurations() const
+{
+  return m_config_descriptor;
+}
+
+std::vector<InterfaceDescriptor> SkylanderUSB::GetInterfaces(u8 config) const
+{
+  return m_interface_descriptor;
+}
+
+std::vector<EndpointDescriptor> SkylanderUSB::GetEndpoints(u8 config, u8 interface, u8 alt) const
+{
+  return m_endpoint_descriptor;
+}
+
+bool SkylanderUSB::Attach()
+{
+  if (m_device_attached)
+  {
+    return true;
+  }
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x}] Opening device", m_vid, m_pid);
+  m_device_attached = true;
+  return true;
+}
+
+bool SkylanderUSB::AttachAndChangeInterface(const u8 interface)
+{
+  return true;
+}
```

Comment:
What parts are still confusing for you? I'd be more than happy to try write them in a way that is less confusing - otherwise yes that method is definitely wrong, let me fix it now

Reply:
The code flow in `SubmitTransfer()`. I think it would be a lot clearer if you had a completely separate branch for the `cmd->request == 0x09` case since it takes up over half the function and the error checking is interleaved with code for other cases.

Reply:
You also seem to call `u8* buf = memory.GetPointerForRange(cmd->data_address, cmd->length);` unconditionally when the result is only relevant in the `cmd->request == 0x09` case...

Reply:
By another branch - do you mean another method within the SkylanderUSB class to handle the 0x09 case?

Reply:
Hold on, give me a few minutes and I'll show you what I mean...

Reply:
https://github.com/dolphin-emu/dolphin/blob/b378ccac889a8a7b42565520d41efcaf978794d5/Source/Core/Core/IOS/USB/Emulated/Skylander.cpp#L110-L417
https://github.com/dolphin-emu/dolphin/commit/b378ccac889a8a7b42565520d41efcaf978794d5

See what I mean? You can reduce the mental complexity a lot by pushing most of the code into a `request == 0x09` branch and exiting early on `cmd->request_type != 0x21`.

Reply:
Ahh okay I see what you mean - exit early when the conditions aren't met, and have the else branch so the bulk of the stuff that actually needs to happen is isolated and easier to read through. On it!

Reply:
Should look a bit better now - added some extra comments in on top of the re-jig. Let me know what you think!